src/hash_pcql.c

```diff
@@ -0,0 +1,322 @@
+	for (Int old_idx = 1; old_idx <= old_capacity; ++old_idx) {
+	    Obj k = ELM_PLIST(old_keys, old_idx);
+	    if (k == 0 || k == Fail)
+	        continue;
+
+        Obj hash = hashfun ? CALL_1ARGS(hashfun, k) : MyHash(k);
+        //if (!IS_INTOBJ(hash))
+        //    ErrorQuit("hashfun failed to return a small int", 0L, 0L);
+
+		// Insert the element from the old table into the new table.
+		// Since we know that no key exists twice in the old table, we
+		// can do this slightly better than by calling lookup, since we
+		// don't have to call _equal().
+
+        Int idx = _PCQL_Hash_Lookup_intern(ht, keys, k, hash, 0, mask, 1, 1);
+
+        Obj v = ELM_PLIST(old_vals, old_idx);
+
+        SET_ELM_PLIST(keys, idx, k);
+        SET_ELM_PLIST(vals, idx, v);
+
+		new_size++;
+    }
+    // Strictly speaking, we should call CHANGED_BAG inside the loop.
+    // However, as we copy elements from existing lists, all the objects
+    // are already known to GASMAN. So it is safe to delay the
+    // notification as no objects can be lost.
+    // FIXME: Verify the above claim!
+    CHANGED_BAG(keys);
+    CHANGED_BAG(vals);
+
+    if (old_size != new_size)
+        ErrorQuit("_PCQL_Hash_Resize_intern: unexpected size change", 0L, 0L);
+
+    // ... and store the result
+    AssPRec(ht, PCQL_Capacity_RNam, INTOBJ_INT(new_capacity));
+    AssPRec(ht, PCQL_Size_RNam, INTOBJ_INT(new_size));
+    AssPRec(ht, PCQL_Deleted_RNam, INTOBJ_INT(0));
+    AssPRec(ht, PCQL_Keys_RNam, keys);
+    AssPRec(ht, PCQL_Vals_RNam, vals);
+}
+
+Obj _PCQL_Hash_Resize_C(Obj self, Obj ht, Obj new_capacity)
+{
+    // TODO: insert extra checks
+/*
+    if new_capacity < ht.size * _PCQL_Hash_Loadfactor then
+        new_capacity := Int( ht.size * _PCQL_Hash_Loadfactor );
+    fi;
+    new_capacity := _PCQL_Hash_RoundUpCapacity( new_capacity );
+*/
+    _PCQL_Hash_Resize_intern(ht, INT_INTOBJ(new_capacity));
+    return 0;
+}
+
+static void _PCQL_GrowIfNecessary(Obj ht)
+{
+    Int capacity = INT_INTOBJ(ElmPRec(ht, PCQL_Capacity_RNam));
+    Int size = INT_INTOBJ(ElmPRec(ht, PCQL_Size_RNam));
+    Int deleted = INT_INTOBJ(ElmPRec(ht, PCQL_Deleted_RNam));
+
+    if ( (size + deleted) * LOADFACTOR_DENOMINATOR > capacity * LOADFACTOR_NUMERATOR) {
+        capacity = 16;
+        while (capacity < 2*size)
+            capacity <<= 1;
+        _PCQL_Hash_Resize_intern(ht, capacity);
+    }
+}
+
+static Obj _PCQL_Hash_SetOrAccValue_C(Obj ht, Obj key, Obj val, int acc)
+{
+    if (key == Fail || val == Fail)
+        ErrorQuit("<key> and <value> must not be equal to 'fail'", 0L, 0L);
+
+    _PCQL_GrowIfNecessary(ht);
+
+    Int idx = INT_INTOBJ(_PCQL_Hash_Lookup(ht, key, 1));
+
+    Obj keys = ElmPRec(ht, PCQL_Keys_RNam);
+    Obj vals = ElmPRec(ht, PCQL_Vals_RNam);
+
+    Obj old_k = ELM_PLIST(keys, idx);
+    if (old_k == Fail) {
+        // we are filling a 'deleted' slot
+        Int deleted = INT_INTOBJ(ElmPRec(ht, PCQL_Deleted_RNam));
+        deleted--;
```

Comment:
Note: orb's hash table collect lots of statistics. While this can dlow you down, it can be super helpful when debugging (for end users of the hashmap, too). So we should also do that, but perhaps optionally. I sense another opportunity to make multiple variations of each function with the help pf templates.